### PR TITLE
refactor: refactor source executor (part 2)

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
This refactor reduces the `target_state` variable passed around. It's hard to understand and error-prone since we already have the fields have similar meanings.

We update the states immediately instead.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15104?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #15103** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15103?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #15104** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15104?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15104?utm_source=stack-comment).

File: src/stream/src/executor/source/state_table_handler.rs [hunk 1]
```diff
@@ -178,14 +178,9 @@ impl<S: StateStore> SourceStateTableHandler<S> {
     where
         SS: SplitMetaData,
     {
-        if states.is_empty() {
-            // TODO should be a clear Error Code
-            bail!("states require not null");
```
Comment (xxchan) [inline src/stream/src/executor/source/state_table_handler.rs:183]: I don't find this check useful?

File: src/stream/src/executor/source/source_executor.rs [hunk 2]
```diff
@@ -284,56 +304,21 @@ impl<S: StateStore> SourceExecutor<S> {
...
-            if should_trim_state && !dropped_splits.is_empty() {
-                // trim dropped splits' state
-                core.split_state_store.trim_state(&dropped_splits).await?;
-            }
-
-            core.latest_split_info = target_splits
-                .into_iter()
-                .map(|split| (split.id(), split))
-                .collect();
-        }
```
Comment (xxchan) [inline src/stream/src/executor/source/source_executor.rs:332]: This section is moved to `update_state_if_changed` now

File: src/stream/src/executor/source/state_table_handler.rs (hunk 1 above)
Comment (tabVersion) [reply src/stream/src/executor/source/state_table_handler.rs:183]: it is ok to remove the check, no need any more

**Comment by xxchan** [comment]
Oops, deterministic test failed

File: src/stream/src/executor/source/source_executor.rs [hunk 3]
```diff
@@ -156,82 +164,96 @@ impl<S: StateStore> SourceExecutor<S> {
...
 
         let mut split_changed = false;
 
         // Checks added splits
-        for (split_id, split) in &target_splits {
-            if let Some(s) = core.updated_splits_in_epoch.get(split_id) {
-                // existing split, no change, clone from cache
-                target_state.push(s.clone())
+        for (split_id, split) in target_splits {
+            if let Some(s) = core.latest_split_info.get(&split_id) {
```
Comment (xxchan) [inline src/stream/src/executor/source/source_executor.rs:198]: This line is a minor change. I think using `latest_split_info` is better than `updated_splits_in_epoch` here.

**Review by tabVersion — APPROVED** [review]
LGTM

**Pull request merged**